File: OpenStreetMap/iOS/LanguageTableViewController.m
//
//  LanguageTableViewController.m
//  Go Map!!
//
//  Created by Bryce on 11/12/15.
//  Copyright © 2015 Bryce. All rights reserved.
//

#import "CommonTagList.h"
#import "LanguageTableViewController.h"



@implementation LanguageTableViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

	_supportedLanguages = [NSMutableArray new];
	NSString * path = [[[NSBundle mainBundle] resourcePath] stringByAppendingPathComponent:@"presets/translations"];
	NSArray * languageFiles = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:path error:NULL];
	languageFiles = [languageFiles arrayByAddingObject:@"en.json"];

	for ( NSString * file in languageFiles ) {
		NSString * code = [file stringByReplacingOccurrencesOfString:@".json" withString:@""];
		NSLocale * locale =  [NSLocale localeWithLocaleIdentifier:code];
		NSString * name = [locale displayNameForKey:NSLocaleIdentifier value:code];
		NSString * name2 = [[NSLocale currentLocale] displayNameForKey:NSLocaleIdentifier value:code];

		[_supportedLanguages addObject:@[ code, name, name2 ]];
	}

	[_supportedLanguages sortUsingComparator:^NSComparisonResult(NSArray * obj1, NSArray * obj2) {
		NSString * s1 = obj1[1];
		NSString * s2 = obj2[1];
		return [s1 compare:s2 options:NSCaseInsensitiveSearch];
	}];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
	if ( section == 0 ) {
		return @"Language selection affects only Presets and only for those presets that are translated for iD. The main interface is still English.";
	}
	return nil;
}


- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
	return _supportedLanguages.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
	UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];
	NSArray * item = _supportedLanguages[ indexPath.row ];

	// name in native language
	NSString * name = item[1];
	cell.textLabel.text = name;

	// name in current language
	cell.detailTextLabel.text = item[2];

	// accessory checkmark
	NSString * code = item[0];
	NSString * preferred = [[NSUserDefaults standardUserDefaults] objectForKey:@"preferredLanguage"];
	cell.accessoryType = [preferred isEqualToString:code] ? UITableViewCellAccessoryCheckmark : UITableViewCellAccessoryNone;
	return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
	NSArray * item = _supportedLanguages[ indexPath.row ];
	NSString * code = item[0];
	[[NSUserDefaults standardUserDefaults] setObject:code forKey:@"preferredLanguage"];

	[self.tableView reloadData];

	[CommonTagList initialize];	// reset tags
}

@end

File: src/iOS/Settings/LanguageTableViewController.swift
//
//  LanguageTableViewController.swift
//  Go Map!!
//
//  Created by Bryce Cogswell on 11/12/15.
//  Copyright © 2015 Bryce Cogswell. All rights reserved.
//

import UIKit

class LanguageTableViewController: UITableViewController {
	override func viewDidLoad() {
		super.viewDidLoad()

		tableView.estimatedRowHeight = 44
		tableView.rowHeight = UITableView.automaticDimension
	}

	// MARK: - Table view data source

	override func numberOfSections(in tableView: UITableView) -> Int {
		return 1
	}

	override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
		if section == 0 {
			let line = NSLocalizedString(
				"Select the language used for Presets. Other app content will use your default language.",
				comment: "")
			return line
		}
		return nil
	}

	override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		return PresetLanguages.languageCodeList.count + 1
	}

	override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)

		var code: String?

		if indexPath.row == 0 {
			// Default
			code = nil

			// name in native language
			cell.textLabel?.text = NSLocalizedString("Automatic", comment: "Automatic selection of presets languages")
			cell.detailTextLabel?.text = nil
		} else {
			code = PresetLanguages.languageCodeList[indexPath.row - 1]

			// name in native language
			cell.textLabel?.text = PresetLanguages.languageNameForCode(code ?? "") ?? ""

			// name in current language
			cell.detailTextLabel?.text = PresetLanguages.localLanguageNameForCode(code ?? "") ?? ""
		}

		// accessory checkmark
		if (PresetLanguages.preferredLanguageIsDefault() && indexPath.row == 0)
			|| code == PresetLanguages.preferredPresetLanguageCode()
		{
			cell.accessoryType = .checkmark
		} else {
			cell.accessoryType = .none
		}
		return cell
	}

	override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
		if indexPath.row == 0 {
			PresetLanguages.setPreferredLanguageCode(nil)
		} else {
			let code = PresetLanguages.languageCodeList[indexPath.row - 1]
			PresetLanguages.setPreferredLanguageCode(code)
		}

		self.tableView.reloadData()

		try? PresetsDatabase.reload(withLanguageCode: PresetLanguages.preferredPresetLanguageCode()) // reset tags
		AppDelegate.shared.mapView.refreshPushpinText()
	}
}
